### ISSUE TITLE
systemd unit file doesn't work

### ISSUE THREAD

**Issue by jbash** [taw00/ipfs-rpm#2]
Thank you for packaging this.

I found quite a few issues with the unit file and associated stuff, several of which kept it from working at all. I've attached a unit file that at least gets the daemon started.

Here's a list of things I found:

Unit file structure
-------------------

### Critical for it to work at all

+ You declare the service as "forking", but ipfs doesn't fork. A while after it's started, systemd will give up on waiting for it to fork and kill the service.
  
  You seem to be trying to deal with this by setting the start timeout to 3 minutes, but it will still always end up killing the process when the 3 minutes are up.
  
  The type should be "exec" or maybe "simple". Probably "exec" because it listens on ports.
  
+ You can't put "IPFS_PATH=$REPO" on the command line in the unit file, because that "command" isn't actually passed to a shell. You have to set the variable either in the environment file or with "Environment=". I don't think either one of those supports copying the value from another variable.
  
  If I were you, I would just eliminate all use of the name "REPO" and set "IPFS_PATH" in the environment file. I find it confusing to have two names for the same thing.
  
+ It looks like you at some point you renamed the scripts in /etc/sysconfig/ipfs, but didn't fix all of the references in the unit file. The stop scripts don't exist. (But I'd just delete the stop script entries anyway; see below).
  
### Recommendations
  
+ A bunch of stuff that's set in the environment file doesn't seem to be used and is therefore confusing. I would suggest just eliminating every variable except for IPFS_PATH, MOUNTPOINT_IPFS, MOUNTPOINT_IPNS (see below), and AUTO_INIT.

+ The start and stop messages from "write-to-journal.sh" are redundant with standard system logging. Echoing "Starting." and "Shutting down" is again redundant. ipfsd itself is also pretty chatty.
  
  The email support is also redundant with stuff you can easily set up using regular system monitoring tools.
  
  You could save yourself work by eliminating all of that stuff from the unit file and the package.
  
+ Any scripts that remain should be able to log to the journal under the unit's name just by printing to stderr. There's no need to pass around a JOURNAL_IDENTIFIER variable.
  
+ I don't think it's ever a good idea to put sleeps into a unit file. It won't permanently wedge the system, but it *will* confuse systemd and slow things down.
  
  systemd will monitor the actual ipfs process. It won't consider it to be shut down, nor shut down any dependent units, nor ever try to start another instance, until it either actually exits, or fails to exit after being given the configured timeout. The timeout you have is pretty short, by the way.
  
+ From I get out of your comments, the StartLimit values you're configuring aren't necessary and don't do what you think they do.

  Systemd will never start more than one instance of any service at the same time, unless you explicitly declare it as a multi-instance service. Even then you have to use special syntax to start a new instance. Typing multiple "systemctl start" commands won't do anything to a service that's already running (assuming the service is properly set up so that systemd can detect when the program has actually exited).
  
+ You have PrivateTmp. There are a ton of other namespacing and sandboxing options, and you might want to look at them sometime. Just be aware that you have to understand Linux namespaces to really know what some of them will do.

### Comments
  
+ You have comments in the unit file that seem to come from the template you used to create it. The problem is that comments originally directed at *you* will confuse novice users who read those comments as if directed at *them*. I would remove that stuff.
  
+ The approach you suggest for multiple instances isn't actually how you'd want to do multiple instances under systemd, so I just wouldn't talk about that.
  
+ A well-behaved Unix/Linux program should always shut down gracefully on SIGTERM.  That's doubly true for anything that might be used as a daemon.  If ipfs needs RPCs to shut down cleanly, that's a bug in ipfs.
  
IPFS arguments
--------------
+ ipfs doesn't understand "--mount-ipfn". Did you mean "--mount-ipns"? The string "ipfn" appears in a couple of other places, too.
  
+ It would be nice if it were possible to start the daemon *without* the FUSE mounts. I think they're probably a security hole, for one thing. Maybe they could be separate services, or controlled by options set up in the environment file?

Working unit file
-----------------
````
# Note: You should probably try to make configuration changes by
# editing /etc/sysconfig/ipfsd instead of by directly modifying this unit
# file.

[Unit]
Description=ipfs daemon
After=syslog.target network.target

[Service]
Type=exec
User=ipfs
Group=ipfs
EnvironmentFile=/etc/sysconfig/ipfsd

# Make sure repository and mount points are set up before daemon starts
ExecStartPre=-/etc/sysconfig/ipfsd-scripts/ipfsd-init.sh $JOURNAL_IDENTIFIER $AUTO_INIT $IPFS_PATH $MOUNTPOINT_IPFS $MOUNTPOINT_IPNS

ExecStart=/usr/bin/ipfs daemon --mount --mount-ipfs=${MOUNTPOINT_IPFS} --mount-ipns=${MOUNTPOINT_IPNS}

# Allow 60 seconds for graceful shutdown after a SIGTERM before resorting
# to strong measures.
TimeoutStopSec=90

# If something triggers an auto-restart, let's wait a bit before taking further action
Restart=on-failure
RestartSec=10

PrivateTmp=true

[Install]
WantedBy=multi-user.target
````

**Comment by taw00** [comment]
Thanks. I will take a look at this as soon as I can.

**Comment by mochaaP** [comment]
@taw00 A possible workaround is `/usr/bin/env VAR=... /usr/bin/ipfs ...`. :)
